Added NoJump on-the-fly example to MSD documentation (Issue #4169)

tani-dubey opened the pull request:
This PR updates the MSD documentation inside `MDAnalysis.analysis.msd` to show users how to correctly apply the NoJump transformation using on-the-fly trajectory transformations.

Issue #4169 noted that although NoJump exists, there was no example showing how to use it directly in MDAnalysis.

This PR adds:
- A minimal working example demonstrating NoJump usage
- Clarification that wrapped coordinates must be unwrapped
- A modern MDAnalysis-based workflow that does not require external tools

Fixes: #4169
I would appreciate any feedback or suggestions for improvement.
Reviews are very welcome, and I’m happy to revise the PR based on your guidance.

<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5165.org.readthedocs.build/en/5165/

<!-- readthedocs-preview mdanalysis end -->

IAlibay commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -27,7 +27,7 @@
 :Authors: Hugo MacDermott-Opeskin
 :Year: 2020
-:Copyright: Lesser GNU Public License v2.1+
+:Copyright: Lesser GNU Public License v2.1
```
```suggestion
:Copyright: Lesser GNU Public License v2.1+
```
Please don't change our license terms.

IAlibay reviewed (dismissed):
Not reviewing the code but enforcing license - please don't change the license.

tani-dubey commented:
Thank you for pointing this out! I’ve updated the file to restore the correct license header (`LGPL v2.1+`).

orbeckst commented on package/AUTHORS:
```diff
@@ -265,6 +265,7 @@ Chronological list of authors
   - Raúl Lois-Cuns  
   - Pranay Pelapkar
   - Shreejan Dolai
+  - **Tanisha Dubey (https://github.com/tanii1125)**
```
follow existing formatting, please

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -65,12 +65,32 @@
+       # Now the trajectory is unwrapped and MSD can be computed normally:
+       from MDAnalysis.analysis.msd import EinsteinMSD
+       MSD = EinsteinMSD(u, select="all", msd_type="xyz")
+       MSD.run()
```
Please run code and post images of the output graph. Demonstrate to the reviewers that the example does what you say it is doing.

When you run code, show us the code that you're running, including the input files.

orbeckst reviewed (changes requested):
Please demonstrate to yourself and your reviewers that your doc update is correct.

tani-dubey commented:
@orbeckst

- Added the example script used to verify the documentation change.
- Updated the AUTHORS file to follow the existing formatting.
- Updated the documentation to clarify that `NoJump` requires periodic
  boundary conditions, avoiding errors when PBC information is not present.

As input , I used the built-in MDAnalysis random walk test data:

- `RANDOM_WALK_TOPO`
- `RANDOM_WALK`

Below is the MSD plot produced by running the example script.
After applying `NoJump`, the MSD increases linearly with lag time, as
expected for normal diffusion, and follows the known 3D Brownian
behavior (≈ 6τ).

<img width="810" height="586" alt="image" src="https://github.com/user-attachments/assets/e4cd949c-83b9-452e-91a7-c5972ae81c59" />

tani-dubey committed:
Reupload commits due to cli fail

orbeckst commented on example/msd_nojump_demo.py:
Thanks for showing the code. However, we do not add example scripts to this repository. Please remove it. Just copy and paste the code in the PR comments for review.

tani-dubey replied:
Thanks for clarifying. I’ll remove the example script from the repository and instead include the code directly in the PR comments for review.

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -69,8 +69,33 @@
    In MDAnalysis you can use the 
-   :class:`~MDAnalysis.transformations.nojump.NoJump`
-   transformation. 
+    :class:`~MDAnalysis.transformations.nojump.NoJump`
```
The indentation looks off. Please correct.

orbeckst reviewed (changes requested):
Is RANDOM_WALK a good example, namely, did you check that it was actually generated with periodic boundary conditions?

In order to demonstrate that the example is working, I'd expect to see a comparison between omitting and using NoJump.

tani-dubey commented:
> Is RANDOM_WALK a good example, namely, did you check that it was actually generated with periodic boundary conditions?
> 
> In order to demonstrate that the example is working, I'd expect to see a comparison between omitting and using NoJump.

Thanks for the clarification,

After checking more carefully, I realized that:-
*  Relying on existing trajectories (e.g. RANDOM_WALK or other test data) is not reliable for demonstrating NoJump, since many of them are either not generated with periodic boundary conditions or do not exhibit explicit single-frame boundary crossings.

Instead of assuming wrapping is present in the input data, I will construct a deterministic test where a periodic box is defined and an artificial jump across the boundary is introduced between consecutive frames. This allows us to verify the intended behavior directly.

tani-dubey committed:
Corrected indentation

tani-dubey commented:
@orbeckst 
Without NoJump, periodic wrapping causes an artificial discontinuity; applying NoJump restores the continuous trajectory across the boundary.
<img width="640" height="480" alt="Figure_1" src="https://github.com/user-attachments/assets/7a91699c-913d-4b4a-9cbf-529ed6d00f63" />

**the code i used to demonstrate this --**

```text
import numpy as np
import matplotlib.pyplot as plt
import MDAnalysis as mda
from MDAnalysis.coordinates.memory import MemoryReader
from MDAnalysis.transformations import NoJump


n_atoms = 1
n_frames = 2
box_length = 10.0  # cubic box, Å

pos0 = np.array([[9.8, 5.0, 5.0]])

pos1 = np.array([[0.2, 5.0, 5.0]])

coords = np.array([pos0, pos1])  # shape (n_frames, n_atoms, 3)

dimensions = np.array([
    [box_length, box_length, box_length, 90, 90, 90],
    [box_length, box_length, box_length, 90, 90, 90],
])

# ----------------------------
# 2. Universe WITHOUT NoJump
# ----------------------------

u_plain = mda.Universe.empty(
    n_atoms,
    trajectory=True,
)

u_plain.load_new(
    coords,
    format=MemoryReader,
    dimensions=dimensions,
)

x_plain = [u_plain.atoms.positions[0,0] for _ in u_plain.trajectory]

# ----------------------------
# 3. Universe WITH NoJump
# ----------------------------

u_nj = mda.Universe.empty(
    n_atoms,
    trajectory=True,
)

u_nj.load_new(
    coords,
    format=MemoryReader,
    dimensions=dimensions,
)

u_nj.trajectory.add_transformations(NoJump())

x_nj = [u_nj.atoms.positions[0,0] for _ in u_nj.trajectory]

# ----------------------------
# 4. Plot
# ----------------------------

frames=[0,1]

plt.figure()
plt.plot(frames, x_plain, marker='o', label="Without nojump")
plt.plot(frames, x_nj,marker='o', label="With nojump")

plt.axhline(box_length, linestyle='--', linewidth=1)
plt.axhline(0, linestyle='--', linewidth=1)

plt.xlabel("Frame")
plt.ylabel("x position (A)")
plt.title("Effect of NoJump on Periodic Boundary Crossing")
plt.legend()
plt.show()

```

orbeckst committed:
Merge branch 'develop' into fix-msd-nojump-doc

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -70,7 +70,31 @@
+        
+        # Apply NoJump transformation to unwrap coordinates
+        nojump = NoJump(u)
+        u.trajectory.add_transformations(nojump)
```
I'd write it more succinctly as
```suggestion
        u.trajectory.add_transformations(NoJump(u))
```

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -70,7 +70,31 @@
+    
+   This example assumes that the trajectory contains periodic box
+   dimensions. If no periodic boundary information is present, box
+   dimensions must be defined before applying ``NoJump``.
```
```suggestion
   dimensions must be defined before applying ``NoJump``, which can
   be accomplished by applying the
   :class:`~MDAnalysis.transformations.boxdimensions.set_dimensions`
   transformation *before* the 
   :class:`~MDAnalysis.transformations.nojump.NoJump` transformation.
```

orbeckst replied:
Can you please check with your example that applying 
```python
transformations = [setdimensions([box_length, box_length, box_length, 90, 90, 90]), NoJump(u)]
```
to a Universe without dimensions works as well?

orbeckst reviewed (changes requested):
Thank you for testing explicitly. This looks good.

I added a note on `setdimensions` (see comments). Could you please check that this works as intended (post code and results). Thank you!

tani-dubey committed:
Update package/MDAnalysis/analysis/msd.py

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

tani-dubey committed:
Update package/MDAnalysis/analysis/msd.py

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

tani-dubey commented:
<img width="640" height="480" alt="example" src="https://github.com/user-attachments/assets/e1219a34-431b-415f-9cdc-d5c45c68da70" />

I verified this workflow with a minimal in-memory example starting from a Universe without initial box dimensions. Applying
`set_dimensions([...])` followed by `NoJump(u)` works as intended and correctly unwraps the coordinates.

Code that I used to verify behaviour-

```python
import numpy as np
import matplotlib.pyplot as plt
import MDAnalysis as mda
from MDAnalysis.coordinates.memory import MemoryReader
from MDAnalysis.transformations import NoJump
from MDAnalysis.transformations.boxdimensions import set_dimensions


n_atoms = 1
n_frames = 2
box_length = 10.0  # cubic box, Å

pos0 = np.array([[9.8, 5.0, 5.0]])

pos1 = np.array([[0.2, 5.0, 5.0]])

coords = np.array([pos0, pos1])  # shape (n_frames, n_atoms, 3)

u_nj = mda.Universe.empty(n_atoms, trajectory=True)

u_nj.load_new(
    coords,
    format=MemoryReader,
)  # NO dimensions 

# transformations
u_nj.trajectory.add_transformations(
    set_dimensions([box_length, box_length, box_length, 90, 90, 90]),
    NoJump(u_nj),
)
x_nj = []
for ts in u_nj.trajectory:
    x_nj.append(u_nj.atoms.positions[0, 0])

# ----------------------------
# 2. Universe WITHOUT NoJump
# ----------------------------

u_plain = mda.Universe.empty(
    n_atoms,
    trajectory=True,
)

u_plain.load_new(
    coords,
    format=MemoryReader,
) 

u_plain.trajectory.add_transformations(
    set_dimensions([box_length, box_length, box_length, 90, 90, 90]),
)

x_plain = []
for ts in u_plain.trajectory:
    x_plain.append(u_plain.atoms.positions[0, 0])

# ----------------------------
# 4. Plot
# ----------------------------

frames = list(range(len(x_plain)))

plt.figure()
plt.plot(frames, x_plain, marker='o', label="set_dim only")
plt.plot(frames, x_nj,marker='o', label="set_dim With nojump")


plt.xlabel("Frame")
plt.ylabel("x position (A)")
plt.title("Effect of NoJump on Periodic Boundary Crossing")
plt.legend()
plt.show()

```

orbeckst reviewed (changes requested):
Thanks for checking that set_dimensions, NoJump is working in a simple example _without_ `MSD`.

You haven't demonstrated yet that it works _with_ `MSD` but you should now know enough about transformations and PBC to create a proper test case. 

My recommendation would be to take the existing `RANDOM_WALK` trajectory and wrap it in a small box (figure out reasonable dimensions by analyzing the random walk) using [wrap](https://docs.mdanalysis.org/stable/documentation_pages/transformations/wrap.html) and write it to a new file. Then use this file as input for MSD (1) with NoJump, (2) without NoJump and compare as before.

tani-dubey commented:
Got it, I used `RANDOM_WALK` and wrapped it into a small periodic box and wrote the resulting coordinates to `random_walk_wrapped.xtc`. This file therefore contains wrapped coordinates and represents a realistic user input trajectory.
Then I used it to demonstrate plot: 1)Without `NoJump` 2)With `NoJump`.
Here is the code I used and the resulting plot -

Code-
```python
import numpy as np
import matplotlib.pyplot as plt
import MDAnalysis as mda

from MDAnalysis.tests.datafiles import RANDOM_WALK, RANDOM_WALK_TOPO
from MDAnalysis.transformations import NoJump, wrap
from MDAnalysis.transformations.boxdimensions import set_dimensions
from MDAnalysis.analysis.msd import EinsteinMSD
from MDAnalysis.coordinates.XTC import XTCWriter


# ----------------------------
# create a input file
# ----------------------------

box = [20.0, 20.0, 20.0, 90.0, 90.0, 90.0]
u = mda.Universe(RANDOM_WALK_TOPO, RANDOM_WALK)

u.trajectory.add_transformations(
    set_dimensions(box),
    wrap(u.atoms),
)

with XTCWriter("random_walk_wrapped.xtc",u.atoms.n_atoms) as W:
    for ts in u.trajectory:
        W.write(u.atoms)

# # ----------------------------
# # 1. MSD WITHOUT NoJump
# # ----------------------------

u= mda.Universe(RANDOM_WALK_TOPO,"random_walk_wrapped.xtc")

msd_wrapped = EinsteinMSD(
    u,
    select="all",
    msd_type="xyz",
    fft=False,
)

msd_wrapped.run()

t = msd_wrapped.results.delta_t_values
msd_nojump = msd_wrapped.results.timeseries.copy()

# # ----------------------------
# # 2. MSD WITH NoJump
# # ----------------------------

# # Reload universe to avoid transformation stacking
u = mda.Universe(RANDOM_WALK_TOPO, "random_walk_wrapped.xtc")

u.trajectory.add_transformations(
    NoJump(u),
)

msd_unwrapped = EinsteinMSD(
    u,
    select="all",
    msd_type="xyz",
    fft=False,
)
msd_unwrapped.run()

msd_with_nojump = msd_unwrapped.results.timeseries

# # ----------------------------
# # 3. Plot comparison
# # ----------------------------

plt.figure(figsize=(6, 4))
plt.plot(t, msd_nojump, label="Wrapped (no NoJump)")
plt.plot(t, msd_with_nojump, label="Unwrapped (with NoJump)")
plt.xlabel("Lag time")
plt.ylabel("MSD")
plt.legend()
plt.title("Effect of NoJump on MSD for RANDOM_WALK")
plt.tight_layout()
plt.show()
```
Output--
<img width="600" height="400" alt="to_give" src="https://github.com/user-attachments/assets/d785c9f2-cc4f-47cd-82b6-2c1a1388c8ec" />

orbeckst commented on package/MDAnalysis/analysis/msd.py:
```diff
@@ -70,7 +70,34 @@
+        from MDAnalysis.analysis.msd import EinsteinMSD
+        MSD = EinsteinMSD(u, select="all", msd_type="xyz")
+        MSD.run()
+    
```
indentation
```suggestion
   A minimal example:

   .. code-block:: python
      
      import MDAnalysis as mda
      from MDAnalysis.transformations import NoJump
      
      u = mda.Universe(TOP, TRAJ)
      
      # Apply NoJump transformation to unwrap coordinates
      u.trajectory.add_transformations(NoJump(u))
       
      # Now the trajectory is unwrapped and MSD can be computed normally:
      from MDAnalysis.analysis.msd import EinsteinMSD
      MSD = EinsteinMSD(u, select="all", msd_type="xyz")
      MSD.run()

```

orbeckst reviewed (approved):
Thank you for checking. This looks good!

I'll fix an indentation problem and then will merge.

orbeckst committed:
Update package/MDAnalysis/analysis/msd.py

orbeckst commented:
@IAlibay do you want to have a look here or can I just dismiss your review as stale? There are no changes to the license in the PR.

orbeckst commented:
@tanii1125 is your name already in the AUTHORS file? If not, please add it.

IAlibay commented:
Please discard my review, thanks!

tani-dubey commented:
@orbeckst, yes my name is in  AUTHORS file

orbeckst commented:
Thank you for your contribution @tanii1125 !
